bson/bsoncodec/empty_interface_codec.go

```diff
@@ -56,24 +52,42 @@ func (eic EmptyInterfaceCodec) EncodeValue(ec EncodeContext, vw bsonrw.ValueWrit
 	return encoder.EncodeValue(ec, vw, val.Elem())
 }
 
+func (eic EmptyInterfaceCodec) getEmptyInterfaceDecodeType(dc DecodeContext, valueType bsontype.Type) (reflect.Type, error) {
+	isDocument := valueType == bsontype.Type(0) || valueType == bsontype.EmbeddedDocument
+	if isDocument && dc.Ancestor != nil {
+		// Using ancestor information rather than looking up the type map entry forces consistent decoding.
+		// If we're decoding into a bson.D, subdocuments should also be decoded as bson.D, even if a type map entry
+		// has been registered.
+		return dc.Ancestor, nil
+	}
+
+	lookupType := valueType
+	if isDocument {
```

Comment:
I wasn't completely sure about this. I think it's fine to treat top-level and embedded documents as the same here. However, this won't work if a user were to register a type map entry for `bsontype.Type(0)` because we only look for `bsontype.EmbeddedDocument`. Given that a type map lookup is cheap, maybe it'd be better to look for `bsontype.EmbeddedDocument` followed by `bsontype.Type(0)` as a fallback?

Reply:
It seems like we just lookup whichever type they gave us first, and fallback to the other one if it wasn't found. So if they gave us Type(0), that is our first lookup, then fallback to EmbeddedDocument. Alternatively, if we got EmbeddedDocument and didn't find it, then fallback to Type(0).

Reply:
Added the fallback code.